<a href="https://colab.research.google.com/github/mdelvecchio2019/Stats2023fall/blob/main/Assignment_3_Delvecchio_Maddy_Stats2023.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Stats Notebook Lab #3

In [2]:
import numpy as np
import matplotlib.pyplot as plt
import imageio
from PIL import Image

we are importing different functions to allow our code to work

### Loops

In [ ]:
for i in range(100):
    print(i)  


0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99


We are assigning the range of 0-100 to the variable i, then we are printing i which shows us the range of numbers 

In [ ]:
for i in range(5,8):
    print(i)

5
6
7


We are assigning the range of 5-8 to the variable i, then we are printing i which shows us the range of numbers 

In [ ]:
for i in range(3):
    print( "Test Number: " , i )

Test Number:  0
Test Number:  1
Test Number:  2


We are assigning the range of 3 to the variable i, then we are printing i as test numbers which shows us the range of numbers as test numbers

### Functions

In [ ]:
def square(x):
    return x**2

We're defining square as x, when a number is ran as x we are returning the value as the number sqaured 

In [ ]:
square(5)

25

We are inputting 5 as x which returns as 5 squared which equals 25

In [ ]:
square(square(2))

16

we are comibing sqaure into a function so we are squaring 2 to get 4 and then squaring 4 to get 16

In [ ]:
square(square(square(2)))


256

we are comibing sqaure into a function so we are squaring 2 to get 4 and then squaring 4 to get 16 then squaring 16 to 256

In [ ]:
square(123456)

15241383936

we are inputting 123456 and it returns as 123456 squared equals 15241383936

In [ ]:
square(square(5))

625

we are comibing sqaure into a function so we are squaring 5 to get 25 and then squaring 25 to get 625

In [ ]:
def add_two(x):
    return x+2

We're defining add_two as x, when a number is ran as x we are returning the value as the number plus 2 

In [ ]:
add_two(5)

7

We are inputting 5 as x which returns as 5 plus 2 which equals 7 

In [ ]:
square(add_two(5))

49

we are combining both the square and add two functions, so we are taking 5 plus 2 to equal 7 and then squaring 7 to get 49

### Logical Branching

In [ ]:
choice = "apple"

we are assigning apple to the variable choice

In [ ]:
if choice == "apple":

    print("You choose apples.")

else:
    
    print("You did not choose apples.")

You choose apples.


we are doing an if else statement, if the choice is equal to apples then we will state that you choose apples if else then we will state you did not choose apples 

In [ ]:
choice = "orange"

we are assigning orange to the variable choice 

In [ ]:
if choice == "apple":

    print("You choose apples.")

else:
    
    print("You did not choose apples.")

You did not choose apples.


we are doing an if else statement, if the choice is equal to apples then we will state that you choose apples if else then we will state you did not choose apples 

In [ ]:
choice = "apple"

we are assigning apple to the variable choice

In [ ]:
if choice == "apple":

    print("You choose apples.")
    print("Thank you, shop again!")
    print(2**123)

elif choice == "orange":

    print("You choose oranges.")
    print("Thank you, shop again!")
    print(2**102)

else:
    
    print("You did not choose apples or oranges.")
    print("Thank you, shop again!")

You choose apples.
Thank you, shop again!
10633823966279326983230456482242756608


we are doing an if else statement, if the choice is equal to apples then we will state that you choose apples, the cost, and Thank you, shop again!, & if the choice is equal to oranges then we will state that you choose oranges, the cost, and Thank you, shop again! if else then we will state you did not choose apples or oranges and Thank you, shop again!

### User Input

In [ ]:
choice = input("Please enter your choice (i.e. apple): ")

Please enter your choice (i.e. apple): pineapple


here we are assigning the user input to the variable choice, with this code we print "Please enter your choice (i.e. apple): " and prompt the user to input an answer

In [ ]:
choice

'pineapple'

we are assigning the user input (bananas) to the variable choice

In [ ]:
if choice == "apple":

    print("You choose apples.")

elif choice == "oranges":

    print("You choose oranges.")

else:
    
    print("You did not choose apples or oranges.")

You did not choose apples or oranges.


we are doing an if else statement, if the choice is equal to apples then we will state that you choose apples & if the choice is equal to oranges then we will state that you choose oranges, if else then we will state you did not choose apples or oranges

#### Coins

In [ ]:
np.random.random()

0.6320581336548893

we are running a random number based off the code above and printing it 

In [ ]:
np.random.random() > 0.5

True

we are asking if the ranomd number is a value less than 0.5 and if it is true/false

In [ ]:
def flip_coin():
    if np.random.random() > 0.5:
        print("Heads")
    else:
        print("Tails")

we are establishing if the numerical value of the coin flip is larger thatn .5 that we print heads and if else we ill print tails

In [ ]:
flip_coin()

Tails


we are establishing that in this course the flipped coin value was more than 0.5 so we are printing heads

### Graphics Output

In [ ]:
from IPython.core.display import display, HTML

we are importing a display to help when we imput a file/statement

In [ ]:
display(HTML('<h1>Hello, world!</h1>'))

we are displaying the statment "Hello, world!"

In [ ]:
def coin():
  return np.random.randint(2)

we are defining coin, but are leaving the value undefined so when it is returned it can be returned as a random value between 0 and 2 

In [ ]:
coin()

0

we are establishing coin and it is coming back as a value

In [ ]:
display(HTML('<img src="https://random-ize.com/coin-flip/us-quarter/us-quarter-front.jpg">'))

we are displaying an image of a coin by using a link to an image found

In [ ]:
heads = 'https://random-ize.com/coin-flip/us-quarter/us-quarter-front.jpg'
tails = 'https://random-ize.com/coin-flip/us-quarter/us-quarter-back.jpg'

here we are assigning heads to one image found and tails to another image found

In [ ]:
def show_img(img):
  display(HTML('<img src=' + img + '>'))

we are defining show_img as one of the two images, in the second piece of code were a pulling the image source and the image will show based off what is imput (heads or tails) or what the value of the number is

In [ ]:
show_img(heads)

we are showing the image of heads

In [ ]:
show_img(tails)

we are showing the image of tails


In [ ]:
def flip_coin():

    if coin():
        show_img(heads)
    else:
        show_img(tails)

we are defining flip_coin as if the coin comes back as a certain value it will show heads and if it comes back as ay other value it will come back as tails 

In [ ]:
flip_coin()

we are showing the flipped coin based off the prior code of if, else statement

In [ ]:
for i in range(5):
    flip_coin()

we are establishing that i is taking on each value from 0-5 and that will flip the coin 5 times, this shows the flipped coin 5 times we are looping the code

In [ ]:
#@title Coin Flip

if coin():

  show_img(heads)

else:

  show_img(tails)

we are making the statment as if the coin comes back as a certain value it will show heads and if it comes back as ay other value it will come back as tails 

### Cards

In [3]:
def shuffle(x):
    for i in range(len(x)-1,0,-1):
        j = np.random.randint(0,i+1)
        x[i],x[j] = x[j],x[i]
        # print(j,x)
    return x

we are defining shuffle as x, and establishing that i is a range of numbers and that x will be a number that starts at -1, take zero steps, and end at -1, j is a random number that starts at 0 and ends at i plus 1, and then the coordinates will be numbers times i and j, at the end it will return with x, a number.

In [ ]:
a = np.arange(52)

we are assigning the array of numbers 0-52 to the variable a

In [ ]:
shuffle(a)

array([32, 30, 27, 19, 40,  0, 34, 15, 14, 31, 38, 36, 12,  3, 28, 46, 10,
        5, 20, 18, 37, 44, 21,  2,  9, 22, 26, 51, 11, 41, 25, 29, 23,  4,
       16, 45, 43,  6,  7, 48, 17, 39,  1,  8, 49, 42, 33, 13, 35, 47, 50,
       24])

we are shuffling the print of the numbers assigned to variable a

In [ ]:
np.random.permutation(52)

array([22, 49, 40, 10,  9, 30, 33, 16, 36, 42, 44,  7, 29, 46, 20,  5, 50,
       47, 41, 32,  3, 35, 21, 11, 48, 24, 37, 39, 45, 26, 43,  6, 23, 28,
       13,  2, 51, 34,  1, 15, 18, 38, 19, 14,  4,  8, 17, 25, 27,  0, 31,
       12])

we are arranging the print of the numbers assigned to variable a in one of the possible random variations it could be arranged 

### Speed Test

In [ ]:
%%timeit
shuffle(a)

182 µs ± 4.11 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


the timeit function gives us a time measurement of the excecution of the code function, so for the execution of shuffle a we are given the amount of time it took to execute 

In [ ]:
%%timeit
np.random.permutation(52)

the timeit function gives us a time measurement of the excecution of the code function, so for the execution of the random permutation of the numbers 0- 52 we are given the amount of time it took to execute

In [ ]:
np.random.permutation(52)

array([51, 40, 39,  2, 19, 24, 12,  7, 14, 28,  8, 21, 18, 49, 47, 27,  1,
       44, 10, 16, 42,  3, 43, 31, 46, 29, 38,  4, 23,  5, 50, 35, 41, 36,
       11, 13, 34, 15, 32, 26, 45, 20, 48, 30,  9, 17, 33,  0, 22, 25,  6,
       37])

we are arranging the print of the numbers assigned to variable a in one of the possible random variations it could be arranged 

In [ ]:
cards = np.random.permutation(52)

we are arranging an array of the numbers assigned to variable a in one of the possible random variations it could be arranged to the variable cards 

In [ ]:
cards

array([32,  0, 51, 29, 27, 15, 39, 45, 42, 12, 30, 28, 44, 34, 33, 16, 49,
       47,  6, 22, 40, 24, 10, 21, 11, 23, 41, 48,  5, 36, 13, 46, 31, 17,
        7, 14,  1, 26,  2, 43, 25,  4, 35, 18,  8, 37, 50,  9,  3, 38, 20,
       19])

we are showing the values of cards

In [ ]:
cards[0]

32

we are showing the 0 card value as 32 

In [ ]:
cards[1]

0

we are showing the 1st card value as 0 

In [ ]:
cards[2]

51

we are showing the 2nd card value as 51 

<img src='http://www.milefoot.com/math/discrete/counting/images/cards.png'>

we are printing an image found

In [ ]:
np.arange(52).reshape(4,13)

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12],
       [13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25],
       [26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38],
       [39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51]])

this is allowing us to reshape the way the array looks by using the shape of 4 row and 13 numbers in each row

In [ ]:
np.arange(52).reshape(4,13)%13

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12],
       [ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12],
       [ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12],
       [ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12]])

this is allowing us to reshape the way the array looks by using the shape of 4 row and 13 numbers in each row with the modual of 13 for each number in the array 0-52

In [ ]:
np.arange(52).reshape(4,13)//13

array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
       [2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2],
       [3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]])

this is allowing us to reshape the way the array looks by using the shape of 4 row and 13 numbers in each row with the dividson of 13 for each number in the array 0-52

In [ ]:
suits = ['Clubs','Spades','Hearts','Diamonds']
names = ['Ace','Two','Three','Four','Five','Six','Seven','Eight','Nine','Ten','Jack','Queen','King']

we assigning names like clubs, spades, etc to the variable suits and the numbers ace through king to the variable names 

In [ ]:
def card_name(card_number):
    
    suit = card_number//13
    name = card_number%13

    return names[name] + " of " + suits[suit]

we are definine card_sname to equal card name 
the suits will equal the suit divided by 13 and the name will equal the modual of 13 and return as the name of and the suit and number 

In [ ]:
card_name(0)

'Ace of Clubs'

the card name at value 0 is ace of clubs

In [ ]:
card_name(51)

'King of Diamonds'

the card name at 51 is king of diamonds


In [ ]:
cards = np.random.permutation(52)

were giving the cards an random permutation of 0-52

In [ ]:
top_of_deck         = card_name(cards[0])
second_card_in_deck = card_name(cards[1])
last_card_in_deck   = card_name(cards[-1])

we are ordering the cards, top of deck will equal the zero value, the second will equal the 1 value and the last card will equal the -1 value 

In [ ]:
top_of_deck, second_card_in_deck, last_card_in_deck

('Three of Hearts', 'Nine of Diamonds', 'Ten of Hearts')

this is showing based off the random permutation what the frist second and last card is 

In [ ]:
for i in range(5):
    print(card_name(cards[i]))

Four of Spades
Six of Diamonds
Ace of Diamonds
Eight of Clubs
Three of Diamonds


this is looping i 5 times and printing each card based off its assigned random permutation

In [ ]:
for i in range(15):
    print(card_name(cards[i]))

Seven of Hearts
Ace of Clubs
King of Diamonds
Four of Hearts
Two of Hearts
Three of Spades
Ace of Diamonds
Seven of Diamonds
Four of Diamonds
King of Clubs
Five of Hearts
Three of Hearts
Six of Diamonds
Nine of Hearts
Eight of Hearts


this is looping i 15 times and printing each card based off its assigned random permutation

https://www.improvemagic.com/all-playing-cards-names-with-pictures/

In [ ]:
card_url_head = "https://www.improvemagic.com/wp-content/uploads/2020/11/"
card_url_tail = ".png"

assigning an image to the heads and tails of the cards 

In [ ]:
suit = ['k','p','s','l']
card = ['a','2','3','4','5','6','7','8','9','10','j','q','k']

assigning the letters to suits and the numbers to cards 

In [3]:
i = np.random.randint(4)
j = np.random.randint(13)

we are assigning i to the randit of 4 and j to randit of j

In [ ]:
i,j

(0, 11)

giving us the (x,y) values of i and j

In [ ]:
suit[i]

'k'

show us the letter assigned to the variable i

In [ ]:
card[j]

'q'

show us the letter assigned to the variable j

In [ ]:
from IPython.core.display import display, HTML

allowing us to import images 

In [ ]:
def show_img(img):
  display(HTML('<img src=' + img + '>'))

allowing us to define images and how their shown

In [ ]:
img = card_url_head + suit[i] + card[j] + card_url_tail

assigning different values to the image which will show us a specific card

In [ ]:
show_img(img)

showing us the specific card from the code above 

In [ ]:
def card_image(card_number):
    
    i = card_number//13
    j = card_number%13

    return suit[i]+card[j]

we are definine card_image to equal card name the suits will equal the i divided by 13 and the j will equal the modual of 13 and return as the image of and the suit and number

In [ ]:
card_image(3)

'k4'

showing us the value of the of the card image 3 

In [ ]:
img = card_url_head + card_image(3) + card_url_tail

assigning different values to the image which will show us a specific card for card 3

In [ ]:
show_img(img)

showing us the image for card 3

In [ ]:
for i in range(52):
    img = card_url_head + card_image(i) + card_url_tail
    show_img(img)

showing us the images for each card in the array from 0-52 after assigning i to all the values in the array